api-authentication/server/index.js

```diff
@@ -0,0 +1,103 @@
+require('dotenv/config');
+const pg = require('pg');
+const argon2 = require('argon2'); // eslint-disable-line
+const express = require('express');
+const jwt = require('jsonwebtoken'); // eslint-disable-line
+const ClientError = require('./client-error');
+const errorMiddleware = require('./error-middleware');
+
+const db = new pg.Pool({ // eslint-disable-line
+  connectionString: process.env.DATABASE_URL,
+  ssl: {
+    rejectUnauthorized: false
+  }
+});
+
+const app = express();
+
+const jsonMiddleware = express.json();
+
+app.use(jsonMiddleware);
+
+app.post('/api/auth/sign-up', (req, res, next) => {
+  const { username, password } = req.body;
+  if (!username || !password) {
+    throw new ClientError(400, 'username and password are required fields');
+  }
+  argon2
+    .hash(password)
+    .then(hashedPassword => {
+      const sql = `
+        insert into "users" ("username", "hashedPassword")
+        values ($1, $2)
+        returning "userId", "username", "createdAt"
+      `;
+      const params = [username, hashedPassword];
+      return db.query(sql, params);
+    })
+    .then(result => {
+      const [user] = result.rows;
+      res.status(201).json(user);
+    })
+    .catch(err => next(err));
+});
+
+app.post('/api/auth/sign-in', (req, res, next) => {
+  const { username, password } = req.body;
+  if (!username || !password) {
+    throw new ClientError(401, 'invalid login');
+  }
+
+  /* your code starts here */
+  const sql = `
+  select "userId", "hashedPassword"
+  from "users"
+  where "username" = $1`;
+
+  const params = [username];
+  db.query(sql, params)
+    .then(result => {
+      if (!result.rows) {
+        throw new ClientError(401, 'invalid login');
+      }
+      argon2
+        .verify(result.rows[0].hashedPassword, password)
+        .then(match => {
+          if (match === false) {
+            throw new ClientError(401, 'invalid login');
+          }
+          const user = {
+            userId: result.rows[0].userId,
+            username: username
```

Comment:
It is always best and safest to use the data from the database when available over user inputted data so for `username` it would be better to use `result.rows[0].username`.

TIP:
Since you are using `result.rows[0]` in several places it would be best to put that into a variable and use the variable everywhere. It is easier to type and even more efficient for the code because it doesn't have to keep accessing the array.  Something like this right before line 60 (Your `if` statement)
```JAVASCRIPT
const user = result.rows[0];
```